pkg/workspace/creds.go

```diff
@@ -30,54 +31,72 @@ import (
 const PulumiCredentialsPathEnvVar = "PULUMI_CREDENTIALS_PATH"
 
-// GetAccessToken returns an access token underneath a given key.
-func GetAccessToken(key string) (string, error) {
+// GetAccount returns an account underneath a given key.
+func GetAccount(key string) (Account, error) {
 	creds, err := GetStoredCredentials()
 	if err != nil && !os.IsNotExist(err) {
-		return "", err
+		return Account{}, err
 	}
-	if creds.AccessTokens == nil {
-		return "", nil
+	if creds.Accounts == nil {
+		token, ok := creds.AccessTokens[key]
+		if !ok {
+			return Account{}, nil
+		}
+		return Account{AccessToken: token}, nil
 	}
-	return creds.AccessTokens[key], nil
+	return creds.Accounts[key], nil
 }
 
-// DeleteAccessToken deletes an access token underneath the given key.
-func DeleteAccessToken(key string) error {
+// DeleteAccount deletes an account underneath the given key.
+func DeleteAccount(key string) error {
 	creds, err := GetStoredCredentials()
 	if err != nil && !os.IsNotExist(err) {
 		return err
 	}
 	if creds.AccessTokens != nil {
 		delete(creds.AccessTokens, key)
 	}
+	if creds.Accounts != nil {
+		delete(creds.Accounts, key)
+	}
 	if creds.Current == key {
 		creds.Current = ""
 	}
 	return StoreCredentials(creds)
 }
 
-// StoreAccessToken saves the given access token underneath the given key.
-func StoreAccessToken(key string, token string, current bool) error {
+// StoreAccount saves the given account underneath the given key.
+func StoreAccount(key string, account Account, current bool) error {
 	creds, err := GetStoredCredentials()
 	if err != nil && !os.IsNotExist(err) {
 		return err
 	}
 	if creds.AccessTokens == nil {
 		creds.AccessTokens = make(map[string]string)
 	}
-	creds.AccessTokens[key] = token
+	if creds.Accounts == nil {
+		creds.Accounts = make(map[string]Account)
+	}
+	creds.AccessTokens[key], creds.Accounts[key] = account.AccessToken, account
 	if current {
 		creds.Current = key
 	}
 	return StoreCredentials(creds)
 }
 
+// Account holds the information associated with a Pulumi account.
+type Account struct {
+	AccessToken string    `json:"accessToken,omitempty"` // The access token for this account.
+	Username    string    `json:"username,omitempty"`    // The username for this account.
+	LastUse     time.Time `json:"lastUse,omitempty"`     // The last time this token was used.
+}
+
 // Credentials hold the information necessary for authenticating Pulumi Cloud API requests.  It contains
 // a map from the cloud API URL to the associated access token.
 type Credentials struct {
-	Current      string            `json:"current,omitempty"`      // the currently selected key.
-	AccessTokens map[string]string `json:"accessTokens,omitempty"` // a map of arbitrary key strings to tokens.
+	Current      string             `json:"current,omitempty"`      // the currently selected key.
+	AccessTokens map[string]string  `json:"accessTokens,omitempty"` // a map of arbitrary key strings to tokens.
```

Comment:
Do we expect it every to be the case where both `AccessTokens` and `Accounts` are set for the same key? Do we want to migrate callers to only using one over the other? etc.

It seems like if we don't make the expected usage clear, it can lead to subtle bugs. (e.g. we only have the `AccessToken` for a key, but not the `Account` or vice versa.

Reply:
The expectation is that either
1. Only `AccessTokens` is populated, or
2. Both `AccessTokens` and `Accounts` are populated, and there is an entry in `AccessTokens` for every entry in `Accounts` (but maybe not vice-versa).

The implementation attempts to ensure this.